src/server/pkg/storage/chunk/transform.go

```diff
@@ -0,0 +1,201 @@
+package chunk
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/cipher"
+	"encoding/hex"
+	io "io"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"golang.org/x/crypto/chacha20"
+)
+
+type CreateOptions struct {
+	Secret      []byte
+	Compression CompressionAlgo
+}
+
+func Create(ctx context.Context, opts CreateOptions, data []byte, createFunc func(ctx context.Context, data []byte) (ID, error)) (*Ref, error) {
+	buf := make([]byte, len(data))
+	// compression
+	compressAlgo, n, err := compress(opts.Compression, buf, data)
+	if err != nil {
+		return nil, err
+	}
+	buf = buf[:n]
+	// encryption
+	dek := encrypt(opts.Secret, buf, buf)
+	// upload
+	id, err := createFunc(ctx, buf)
+	if err != nil {
+		return nil, err
+	}
+	return &Ref{
+		Id:              id,
+		SizeBytes:       int64(len(buf)),
+		Dek:             dek,
+		CompressionAlgo: compressAlgo,
+	}, nil
+}
+
+func Get(ctx context.Context, memCache obj.Client, ref *Ref, w io.Writer, getFunc func(ctx context.Context, id ID, w io.Writer) error) error {
+	// check cache
+	if err := getFromCache(ctx, memCache, ref, w); err == nil {
+		return nil
+	}
+	buf := &bytes.Buffer{}
+	if err := getFunc(ctx, ref.Id, buf); err != nil {
+		return err
+	}
+	// validation
+	actualHash := Hash(buf.Bytes())
+	if !bytes.Equal(actualHash, ref.Id) {
+		return errors.Errorf("bad chunk. HAVE: %x WANT: %x", actualHash, ref.Id)
+	}
+	// decryption
+	var r io.Reader = buf
+	var err error
+	if r, err = decrypt(ref.Dek, r); err != nil {
+		return err
+	}
+	// decompression
+	if r, err = decompress(ref.CompressionAlgo, r); err != nil {
+		return err
+	}
+	// write to cache
+	if cw, err := cacheWriter(ctx, memCache, ref); err == nil {
+		w = io.MultiWriter(w, cw)
+	}
+	_, err = io.Copy(w, r)
+	return err
+}
+
+// compress attempts to compress src using algo. If the compressed data is bigger
+// then no compression is used.
+// compress returns the compression algorithm used (algo or NONE), the number of bytes written to dst
+// or an error
+func compress(algo CompressionAlgo, dst, src []byte) (CompressionAlgo, int, error) {
+	switch algo {
+	case CompressionAlgo_NONE:
+		copy(dst, src)
+		return CompressionAlgo_NONE, len(src), nil
+	case CompressionAlgo_GZIP_BEST_SPEED:
+		lw := newLimitWriter(dst)
+		err := func() error {
+			gw, err := gzip.NewWriterLevel(lw, gzip.BestSpeed)
+			if err != nil {
+				return err
+			}
+			_, err = gw.Write(src)
+			if err != nil {
+				return err
+			}
+			return gw.Close()
```

Comment:
We probably should defer this. I don't think it really matters in this case, but I think we should always follow that pattern unless it would result with incorrect behavior (like closing the object storage writers).